embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+pub struct Channel<M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    mutex: M,
+    state: ChannelState<T, N>,
+}
+
+pub type WithCriticalSections = CriticalSectionMutex<()>;
+
+impl<T, const N: usize> Channel<WithCriticalSections, T, N> {
+    /// Establish a new bounded channel using critical sections. Critical sections
+    /// should be used only single core targets where communication is required
+    /// from exception mode e.g. interrupt handlers. To create one:
+    ///
+    /// ```
+    /// use embassy::util::mpsc;
+    /// use embassy::util::mpsc::{Channel, WithCriticalSections};
+    ///
+    /// // Declare a bounded channel of 3 u32s.
+    /// let mut channel = Channel::<WithCriticalSections, u32, 3>::with_critical_sections();
+    /// // once we have a channel, obtain its sender and receiver
+    /// let (sender, receiver) = mpsc::split(&channel);
+    /// ```
+    pub const fn with_critical_sections() -> Self {
+        let mutex = CriticalSectionMutex::new(());
+        let state = ChannelState::new();
+        Channel { mutex, state }
+    }
+}
+
+pub type WithThreadModeOnly = ThreadModeMutex<()>;
+
+impl<T, const N: usize> Channel<WithThreadModeOnly, T, N> {
+    /// Establish a new bounded channel for use in Cortex-M thread mode. Thread
+    /// mode is intended for application threads on a single core, not interrupts.
+    /// As such, only one task at a time can acquire a resource and so this
+    /// channel avoids all locks. To create one:
+    ///
+    /// ``` no_run
+    /// use embassy::util::mpsc;
+    /// use embassy::util::mpsc::{Channel, WithThreadModeOnly};
+    ///
+    /// // Declare a bounded channel of 3 u32s.
+    /// let mut channel = Channel::<WithThreadModeOnly, u32, 3>::with_thread_mode_only();
+    /// // once we have a channel, obtain its sender and receiver
+    /// let (sender, receiver) = mpsc::split(&channel);
+    /// ```
+    pub const fn with_thread_mode_only() -> Self {
+        let mutex = ThreadModeMutex::new(());
+        let state = ChannelState::new();
+        Channel { mutex, state }
+    }
+}
+
+pub type WithNoThreads = NoopMutex<()>;
+
+impl<T, const N: usize> Channel<WithNoThreads, T, N> {
+    /// Establish a new bounded channel for within a single thread. To create one:
+    ///
+    /// ```
+    /// use embassy::util::mpsc;
+    /// use embassy::util::mpsc::{Channel, WithNoThreads};
+    ///
+    /// // Declare a bounded channel of 3 u32s.
+    /// let mut channel = Channel::<WithNoThreads, u32, 3>::with_no_threads();
+    /// // once we have a channel, obtain its sender and receiver
+    /// let (sender, receiver) = mpsc::split(&channel);
+    /// ```
+    pub const fn with_no_threads() -> Self {
+        let mutex = NoopMutex::new(());
+        let state = ChannelState::new();
+        Channel { mutex, state }
+    }
+}
+
+impl<M, T, const N: usize> Channel<M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    fn get(&self) -> &mut Self {
```

Comment:
converting `&self` to `&mut self` like this is unsound: you may not have two `&mut` to the same data, or a `&mut` and a `&` to the same data.

Reply:
Yes. This is effectively equivalent to UnsafeCell though, right? Only the senders and receiver may access this method and they cannot outlive the channel. Each of the senders and the one receiver must point back to the channel.

Reply:
Yes and no.  First, if this method requires "special care" to be used correctly to not cause UB, it should be `unsafe fn`.

Second, you may say "okay, it's `unsafe fn`, and Sender/Receiver take care to use it correctly (only call it while holding the mutex and only once, so there's never two `&mut` concurrently), it's *still* UB.

Why?

The compiler is allowed to assume a function taking `&` never mutates the data if the data is not behind UnsafeCell. For example:

```rust
struct Foo { bar: u32 }

let foo = Foo { bar: 42 }
do_stuff(foo);
whatever(foo.bar);
```

The compiler is allowed to optimize the last line to `whatever(42)`, because it's allowed to assume `do_stuff` can't mutate `foo.bar` because it takes a `&Foo`. It is UB for `do_stuff` to use `unsafe` to convert `&Foo` to `&mut Foo`. It can cause Bad Stuff to happen, like broken optimizations as seen here.

`UnsafeCell` is "special", it warns the compiler that "interior mutability" can happen, which will prevent these optimizations *for data inside the `UnsafeCell` ONLY*. ALL mutated data must be inside the `UnsafeCell`. 

Also, it's STILL not allowed to cast `&` to `&mut`. Even if everything in the struct is inside an `UnsafeCell`. You're only allowed to get a `&mut` to stuff *inside the `UnsafeCell`*, not to the outer struct. Castinv `&` to `&mut` is ALWAYS UB, no exceptions.

This is the reason why all types that allow for interior mutability always end up using `UnsafeCell` internally: Cell, RefCell, Mutex...

Reply:
Further reading: 

- https://doc.rust-lang.org/reference/interior-mutability.html
- https://doc.rust-lang.org/reference/behavior-considered-undefined.html

Most notably, from the last link:

> - Breaking the pointer aliasing rules. &mut T and &T follow LLVM’s scoped noalias model, except if the &T contains an UnsafeCell<U>.
> - Mutating immutable data. All data inside a const item is immutable. Moreover, all data reached through a shared reference or data owned by an immutable binding is immutable, unless that data is contained within an UnsafeCell<U>.

Reply:
I never realised that `UnsafeCell` was special to the compiler. Thanks for the detailed explanations. I'm going to overhaul this area and look to use `RefCell` instead.

Reply:
RefCell is not zero-cost though. I think it's possible to get away with just Cell/UnsafeCell.

Reply:
> RefCell is not zero-cost though. I think it's possible to get away with just Cell/UnsafeCell.

Ok

Reply:
I've attended to this via commit 29bfa6b7b3d1a3108382d8b4cec10617c24878cc. I used `UnsafeCell`. `Cell` would require the use of `take` and `swap` to perform the interior mutation, which I don't think makes sense for us here as we want multiple threads to be able to access the internal channel, but locked when required.

To this end, I've introduced a `lock` method on the channel in place of the UB `get` that I had before. This ensures that all access to the channel is synchronised. A new `ChannelCell` has been introduced to support our requirements around interior mutability. The channel's methods have been moved to the `ChannelState` type, which also greatly simplifies the code.